Objective

Understand the structure, quality, and characteristics of the dataset before performing any cleaning or feature engineering.

In [1]:
import pandas as pd
import numpy as np

from pathlib import Path

In [2]:
DATA_PATH = Path("../data/raw/application_train.csv")
df = pd.read_csv(DATA_PATH)

In [3]:
df.shape

(307511, 122)

In [4]:
df.head()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
3,100006,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,...,0,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN
4,100007,0,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0


In [5]:
df.tail()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
307506,456251,0,Cash loans,M,N,N,0,157500.0,254700.0,27558.0,...,0,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN
307507,456252,0,Cash loans,F,N,Y,0,72000.0,269550.0,12001.5,...,0,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN
307508,456253,0,Cash loans,F,N,Y,0,153000.0,677664.0,29979.0,...,0,0,0,0,1.0,0.0,0.0,1.0,0.0,1.0
307509,456254,1,Cash loans,F,N,Y,0,171000.0,370107.0,20205.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
307510,456255,0,Cash loans,F,N,N,0,157500.0,675000.0,49117.5,...,0,0,0,0,0.0,0.0,0.0,2.0,0.0,1.0


In [8]:
df.sample(5)

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
193922,324871,0,Cash loans,M,N,N,0,135000.0,269550.0,18234.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
91420,206165,0,Cash loans,F,Y,Y,1,225000.0,1502941.5,47322.0,...,0,0,0,0,0.0,0.0,0.0,0.0,1.0,4.0
93309,208355,0,Revolving loans,F,N,N,0,98100.0,157500.0,7875.0,...,0,0,0,0,0.0,0.0,0.0,0.0,1.0,8.0
306362,454952,1,Cash loans,F,N,Y,0,69750.0,254700.0,17019.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
71462,182874,0,Cash loans,M,N,Y,0,495000.0,2160000.0,85792.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0


In [9]:
dtype_summary = (
    df.dtypes
      .astype(str)
      .value_counts()
      .reset_index()
)

dtype_summary.columns = ["Data Type", "Count"]

dtype_summary

,Data Type,Count
0,float64,65
1,int64,41
2,object,16


In [10]:
numerical_features = df.select_dtypes(include=["int64", "float64"]).columns.tolist()
categorical_features = df.select_dtypes(include=["object"]).columns.tolist()
boolean_features = df.select_dtypes(include=["bool"]).columns.tolist()

print(f"Numerical Features : {len(numerical_features)}")
print(f"Categorical Features : {len(categorical_features)}")
print(f"Boolean Features : {len(boolean_features)}")

Numerical Features : 106
Categorical Features : 16
Boolean Features : 0


In [11]:
# ==========================================
# Section 6 : Memory Usage
# ==========================================

memory_usage = (
    df.memory_usage(deep=True)
      .sort_values(ascending=False)
      .reset_index()
)

memory_usage.columns = ["Column", "Memory (Bytes)"]

memory_usage["Memory (MB)"] = memory_usage["Memory (Bytes)"] / (1024**2)

memory_usage.head(10)

,Column,Memory (Bytes),Memory (MB)
0,NAME_EDUCATION_TYPE,22833595,21.775813
1,NAME_HOUSING_TYPE,20234150,19.296789
2,ORGANIZATION_TYPE,18919932,18.043453
3,NAME_TYPE_SUITE,18737010,17.869005
4,NAME_INCOME_TYPE,18392725,17.540669
5,NAME_CONTRACT_TYPE,18289544,17.442268
6,NAME_FAMILY_STATUS,18027177,17.192056
7,WEEKDAY_APPR_PROCESS_START,17291694,16.490644
8,OCCUPATION_TYPE,15657022,14.931700
9,CODE_GENDER,15375558,14.663275


In [13]:
# ==========================================
# Section 7 : Duplicate Analysis
# ==========================================

duplicate_rows = df.duplicated().sum()

print(f"Duplicate Rows : {duplicate_rows}")

Duplicate Rows : 0


In [14]:
duplicate_ids = df["SK_ID_CURR"].duplicated().sum()

print(f"Duplicate Customer IDs : {duplicate_ids}")

Duplicate Customer IDs : 0


In [21]:
# ==========================================
# Section 8 : Target Variable Analysis
# ==========================================
target_summary = (
    df["TARGET"]
    .value_counts()
    .rename_axis("TARGET")
    .reset_index(name="Count")
)

target_summary["Percentage"] = (
    target_summary["Count"] / len(df)
) * 100

target_summary

,TARGET,Count,Percentage
0,0,282686,91.927118
1,1,24825,8.072882


In [22]:
majority = target_summary["Count"].max()
minority = target_summary["Count"].min()

print(f"Imbalance Ratio : {majority/minority:.2f} : 1")

Imbalance Ratio : 11.39 : 1


In [23]:
# ==========================================
# Section 9 : Missing Value Analysis
# ==========================================

missing_summary = pd.DataFrame({

    "Feature": df.columns,
    "Missing Count": df.isnull().sum().values,

    "Missing Percentage":
    (df.isnull().sum().values / len(df)) * 100

})

missing_summary = missing_summary.sort_values(
    by="Missing Percentage",
    ascending=False
)

missing_summary.head(20)

,Feature,Missing Count,Missing Percentage
48,COMMONAREA_AVG,214865,69.872297
62,COMMONAREA_MODE,214865,69.872297
76,COMMONAREA_MEDI,214865,69.872297
84,NONLIVINGAPARTMENTS_MEDI,213514,69.432963
70,NONLIVINGAPARTMENTS_MODE,213514,69.432963
56,NONLIVINGAPARTMENTS_AVG,213514,69.432963
86,FONDKAPREMONT_MODE,210295,68.386172
54,LIVINGAPARTMENTS_AVG,210199,68.354953
82,LIVINGAPARTMENTS_MEDI,210199,68.354953
68,LIVINGAPARTMENTS_MODE,210199,68.354953


In [24]:
missing_summary[
    missing_summary["Missing Percentage"] > 50
]

,Feature,Missing Count,Missing Percentage
48,COMMONAREA_AVG,214865,69.872297
62,COMMONAREA_MODE,214865,69.872297
76,COMMONAREA_MEDI,214865,69.872297
84,NONLIVINGAPARTMENTS_MEDI,213514,69.432963
70,NONLIVINGAPARTMENTS_MODE,213514,69.432963
56,NONLIVINGAPARTMENTS_AVG,213514,69.432963
86,FONDKAPREMONT_MODE,210295,68.386172
54,LIVINGAPARTMENTS_AVG,210199,68.354953
82,LIVINGAPARTMENTS_MEDI,210199,68.354953
68,LIVINGAPARTMENTS_MODE,210199,68.354953


In [28]:
feature_categories = {

    "Target": ["TARGET"],

    "Identifiers": [
        "SK_ID_CURR"
    ],

    "Loan Features": [
        col for col in df.columns
        if "AMT_" in col
    ],

    "External Scores": [
        col for col in df.columns
        if "EXT_SOURCE" in col
    ],

    "Document Flags": [
        col for col in df.columns
        if "FLAG_DOCUMENT" in col
    ],

    "Region Features": [
        col for col in df.columns
        if "REGION" in col
    ],

    "Time Features": [
        col for col in df.columns
        if "DAYS" in col
    ]
}
# print(feature_categories)
for category, cols in feature_categories.items():
    print(f"{category} ({len(cols)})")
    print(cols)
    print("-"*60)

Target (1)
['TARGET']
------------------------------------------------------------
Identifiers (1)
['SK_ID_CURR']
------------------------------------------------------------
Loan Features (10)
['AMT_INCOME_TOTAL', 'AMT_CREDIT', 'AMT_ANNUITY', 'AMT_GOODS_PRICE', 'AMT_REQ_CREDIT_BUREAU_HOUR', 'AMT_REQ_CREDIT_BUREAU_DAY', 'AMT_REQ_CREDIT_BUREAU_WEEK', 'AMT_REQ_CREDIT_BUREAU_MON', 'AMT_REQ_CREDIT_BUREAU_QRT', 'AMT_REQ_CREDIT_BUREAU_YEAR']
------------------------------------------------------------
External Scores (3)
['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']
------------------------------------------------------------
Document Flags (20)
['FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3', 'FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6', 'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9', 'FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12', 'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15', 'FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_1

In [29]:
# ==========================================
# Numerical Feature Analysis
# ==========================================

int_features = df.select_dtypes(include="int64").columns.tolist()

float_features = df.select_dtypes(include="float64").columns.tolist()

print(f"Integer Features : {len(int_features)}")

print(f"Float Features : {len(float_features)}")

Integer Features : 41
Float Features : 65


In [30]:
time_features = [
    col for col in df.columns
    if "DAYS" in col
]

time_features

['DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH',
 'DAYS_LAST_PHONE_CHANGE']

In [31]:
cardinality = pd.DataFrame({

    "Feature": categorical_features,

    "Unique Values":
    [df[col].nunique() for col in categorical_features]

})

cardinality.sort_values(
    by="Unique Values",
    ascending=False
)

,Feature,Unique Values
11,ORGANIZATION_TYPE,58
9,OCCUPATION_TYPE,18
5,NAME_INCOME_TYPE,8
4,NAME_TYPE_SUITE,7
14,WALLSMATERIAL_MODE,7
10,WEEKDAY_APPR_PROCESS_START,7
7,NAME_FAMILY_STATUS,6
8,NAME_HOUSING_TYPE,6
6,NAME_EDUCATION_TYPE,5
12,FONDKAPREMONT_MODE,4


In [32]:
binary_features = [
    col
    for col in categorical_features
    if df[col].nunique()==2
]

binary_features

['NAME_CONTRACT_TYPE',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'EMERGENCYSTATE_MODE']

In [33]:
identifier_columns = [

    col
    for col in df.columns
    if "SK_ID" in col
]

identifier_columns

['SK_ID_CURR']

In [35]:
high_cardinality = [

    col

    for col in categorical_features

    if df[col].nunique() > 20

]

high_cardinality

['ORGANIZATION_TYPE']

In [36]:
quality_report = pd.DataFrame({

    "Check":[

        "Missing Values",

        "Duplicate Rows",

        "Duplicate IDs",

        "Target Available",

        "Mixed Data Types",

        "High Cardinality"

    ],

    "Status":[

        "✔️" if df.isnull().sum().sum()>0 else "❌",

        "✔️" if duplicate_rows==0 else "❌",

        "✔️" if duplicate_ids==0 else "❌",

        "✔️" if "TARGET" in df.columns else "❌",

        "✔️",

        "✔️" if len(high_cardinality)>0 else "❌"

    ]

})

quality_report

,Check,Status
0,Missing Values,✔️
1,Duplicate Rows,✔️
2,Duplicate IDs,✔️
3,Target Available,✔️
4,Mixed Data Types,✔️
5,High Cardinality,✔️


## Key Observations

- The dataset contains approximately 307,000 loan applications with 122 features.
- The target variable (`TARGET`) is binary and exhibits class imbalance, so evaluation metrics such as ROC-AUC, Precision, Recall, F1-score, and PR-AUC will be more informative than accuracy.
- Several features contain a high percentage of missing values and will require careful handling during preprocessing.
- The dataset includes both numerical and categorical variables, requiring different preprocessing strategies.
- `SK_ID_CURR` is a unique customer identifier and should not be used directly as a predictive feature. It will later serve as the primary key for joining additional datasets.
- Features such as `EXT_SOURCE_1`, `EXT_SOURCE_2`, and `EXT_SOURCE_3` are expected to be highly predictive and deserve closer examination during EDA.
- Time-related variables (e.g., `DAYS_BIRTH`, `DAYS_EMPLOYED`) are stored as negative day offsets and will need transformation into meaningful units such as years.
- The presence of high-cardinality and near-constant features suggests that feature selection will be an important step before model training.

In [44]:
import sys
from pathlib import Path

# Get project root (parent of notebooks)
project_root = Path.cwd().parent

# Add project root to Python path
sys.path.append(str(project_root))

In [45]:
from  src.utils.data_profiler import profile_dataset

profile = profile_dataset(df)

profile

,Metric,Value
0,Rows,307511.00
1,Columns,122.00
2,Numerical Columns,106.00
3,Categorical Columns,16.00
4,Boolean Columns,0.00
5,Duplicate Rows,0.00
6,Missing Values,9152465.00
7,Memory Usage (MB),504.99


In [10]:
from src.utils.data_profiler import profile_dataset

In [3]:
import sys

print(sys.executable)


c:\Users\RISHAV\AppData\Local\Programs\Python\Python312\python.exe


In [8]:
import os
print(os.getcwd())

d:\AI_Credit_Risk_Analysis_Platform\notebook


In [9]:
from pathlib import Path
import sys

# Get the project root
PROJECT_ROOT = Path.cwd().parent

# Add it to Python's search path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

print(PROJECT_ROOT)

d:\AI_Credit_Risk_Analysis_Platform
